packages/vite/src/node/config.ts

```diff
@@ -376,11 +376,11 @@ export async function resolveConfig(
   )
 
   // call configResolved hooks
-  userPlugins.forEach((p) => {
+  for (const p of userPlugins) {
```

Comment:
Could use `Promise.all` here, as `configResolved` should not change the configure object anyway.

Reply:
@antfu I think it is better to avoid `Promise.all` as all other hooks are sequential, and plugins could set up an internal state that will be used by following plugins. For example for presets where you return a plugin that is an array of plugins. I think the performance gains here may not justify the possible subtle issues. What do you think?

Reply:
I had Promise.all first but didn't want to change the documented sequential behavior.

Reply:
With `Promise.all` they are still sequential for sync functions and async functions before their first `await`. I just thought `configResolved` is designed for plugins to grab the configuration and store it, and that's it. I am not sure if it's ideal to let all the plugins wait for each other for each hook.

Reply:
You have a point, and in Rollup, there are a lot of parallel async hooks so we need to choose here.
I think that it was a good idea to make the `config` hook be sequential async (since the `options` hook in rollup is also sequential https://rollupjs.org/guide/en/#options). But for this one, I think there is no equivalent hook in rollup so we have to decide. I am fine with this one being parallel, we need to document this in [the same way as rollup does](https://rollupjs.org/guide/en/#build-hooks) in the Plugin API section. There is not mention of hooks being sequential or parallel in our docs at this point.

Reply:
the docs do mention sequential, that was the reason i didn't go with Promise.all because i didn't want to change it in case anyone already relies on it. But antfu is right, for sync functions all behaves the same, so i'm indifferent.

Docs: https://github.com/vitejs/vite/pull/2951/files#diff-4f815d9b4fc689178afacaefc7f464c1a2493e0896aad5916f2b025891dfb889R193

My main reasoning for having this hook as async too is that ResolvedConfig passed to configResolved has a lot more information available than UserConfig passed to config. eg `isProduction` and `command`.

Reply:
You are right, sorry, I only looked for `parallel`. I think Evan used `sequential` because it was `sync` here. @antfu if you think we can go ahead with making this one `parallel`, I am fine with it. If not, we should request a review from Evan to check.

Reply:
Yeah, maybe it's better to have Evan's feedback first.

Reply:
I think we could still merge this one as parallel, because later if it is decided to change it to sequential it wouldnt be a breacking change, no? It would be breacking if we changre from sequential to parallel. Or should this be a minor in both cases?

Reply:
rebased main and changed to parallel with Promise.all